x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -1627,7 +1628,7 @@ LogEntryBuilder withAuthentication(Authentication authentication) {
             } else {
                 final Authentication.RealmRef authenticatedBy = authentication.getAuthenticatingSubject().getRealm();
                 if (authentication.isRunAs()) {
-                    final Authentication.RealmRef lookedUpBy = authentication.getLookedUpBy();
+                    final Authentication.RealmRef lookedUpBy = authentication.getEffectiveSubject().getRealm();
```

Comment:
This is a great observation! This turned out to be a bug! When run-as fails (because user does not exist) and authentication_success event is enabled for audit logging, the server throws a NPE

> Cannot invoke \"org.elasticsearch.xpack.core.security.authc.Authentication$RealmRef.getName()\" because \"lookedUpBy\" is null

Since this is an existing bug, I'll have a separate PR to address this. Thanks for raising it.

Reply:
Glad I jotted it down! Thanks for following up and figuring out the NPE. Separate PR makes sense 👍